demo_nodes_py/demo_nodes_py/parameters/async_param_client.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import threading
+
+from ament_index_python import get_package_share_directory
+import rclpy
+import rclpy.context
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.parameter import Parameter
+from rclpy.parameter import parameter_dict_from_yaml_file
+from rclpy.parameter import parameter_value_to_python
+from rclpy.parameter_client import AsyncParameterClient
+
+
+def main(args=None):
+    target_node_name = 'param_test_target'
+
+    context = rclpy.context.Context()
+    rclpy.init(context=context)
+    target_executor = SingleThreadedExecutor(context=context)
+    client_executor = SingleThreadedExecutor(context=context)
+
+    node = rclpy.create_node('async_param_client', context=context)
+    target_node = rclpy.create_node(target_node_name,
+                                    allow_undeclared_parameters=True,
+                                    context=context)
+    target_node.declare_parameters('', [
+        ('int_parameter', 1),
+        ('bool_parameter', False),
+        ('string_parameter', 'Hello World'),
+    ])
+    target_executor.add_node(target_node)
```

Comment:
In the spirit of keeping the demo simple, can we just use a single node (ie. use the same node to create the parameters client and as the target node)? Then we can avoid creating an executor and spinning in a separate thread.

Reply:
My original intent for having two nodes and two seperate contexts is to demonstrate the new functionality of first-class support for setting a remote nodes' parameters. I feel like having only a single node in the demo defeats that purpose since a node setting parameters for itself would presumably just use `self.set_parameters`.

That being said, I agree with keeping the demo simple. I think the best way to do that would be to not exhaustively run through every method offered by `ParameterClient` -- it is a little repetitive and overruns this demo with boilerplate code.

Reply:
I suppose if we really want to demo interacting with a remote node, then that node should be run independently (i.e. in a separate process). E.g. we could add a comment to this demo that the user should first run the [parameter blackboard](https://github.com/ros2/demos/blob/93e3cf89b3d4882b2733c63a008d13672517cab9/demo_nodes_cpp/src/parameters/parameter_blackboard.cpp) executable (or a new python equivalent) as part of this demo.

I'm also still okay with using a single node since this is what the cpp demos do, [for example](https://github.com/ros2/demos/blob/93e3cf89b3d4882b2733c63a008d13672517cab9/demo_nodes_cpp/src/parameters/parameter_blackboard.cpp).

Reply:
Using the parameter blackboard sounds like a good idea, I'll work something out for that